backend-worker/ucm_valuation.py

```diff
@@ -0,0 +1,494 @@
+        Celsius. For now, we simply convert the incoming temperature data to Fahrenheit then convert the output
+        Fahrenheit Degree Day to Celsius.
+
+    """
+    t_air_nodata = pygeoprocessing.get_raster_info(str(t_air_raster_path))["nodata"][0]
+
+    def cdd_op(t_air_array, cdd_val):
+        cdd = np.empty(t_air_array.shape, dtype=np.float32)
+        cdd[:] = TARGET_NODATA
+
+        valid_mask = slice(None)
+        if t_air_nodata is not None:
+            valid_mask = ~np.isclose(t_air_array, t_air_nodata)
+        t_air_valid = t_air_array[valid_mask]
+
+        cdd[valid_mask] = np.where(
+            t_air_valid >= cdd_val,
+            (29.58 * (t_air_valid * 9 / 5 + 32) - 1905) * 5 / 9,
+            (1.07 * (10 ** (-18)) * ((t_air_valid * 9 / 5 + 32) ** 10.96)) * 5 / 9,
+        )
+        return cdd
+
+    pygeoprocessing.raster_calculator(
+        [(str(t_air_raster_path), 1), (cdd_threshold, "raw")],
+        cdd_op,
+        str(target_cdd_path),
+        gdal.GDT_Float32,
+        TARGET_NODATA,
+    )
+
+
+def mortality_risk_calculation(
+    t_air_raster_path: pathlike,
+    target_mortality_path: pathlike,
+    mortality_risk_df: pd.DataFrame,
+) -> None:
+    """Raster calculator op to calculate Relative Mortality Risk based on the following function:
+    mortality = (Ti - T0) / (T0 - T1) * (R0 - R1) + R1
+        where:
+            Ti = T air raster
+            T0 = upper temperature threshold
+            T1 = lower temperature threshold
+            R0 = upper mortality risk
+            R1 = lower mortality risk
+
+    Args:
+        t_air_raster_path (pathlike): Path to T air raster.
+        target_mortality_path (pathlike): Path to target mortality risk raster.
+        mortality_risk_df (DataFrame): Pandas DataFrame with columns for the temperature thresholds and associated
+            mortality risk
+
+    Returns:
+        None
+
+    """
+    # Ensure path variables are Path objects
+    t_air_raster_path = Path(t_air_raster_path)
+    target_mortality_path = Path(target_mortality_path)
+
+    t_air_nodata = pygeoprocessing.get_raster_info(str(t_air_raster_path))["nodata"][0]
+
+    def mortality_op(t_air_array):
+        """
+        Raster calculator op that calculates relative mortality risk based on temperature thresholds.
+        mortality = (Ti - T0) / (T0 - T1) * (R0 - R1) + R1
+        """
+        mortality = np.empty(t_air_array.shape, dtype=np.float32)
+        mortality[:] = TARGET_NODATA
+
+        if t_air_nodata is not None:
+            valid_mask = ~np.isclose(t_air_array, t_air_nodata)
+        else:
+            valid_mask = np.ones(t_air_array.shape, dtype=bool)
+
+        thresholds = ["01", "10", "mmtp", "90", "99"]
+        # Iterate through thresholds, except 99th percentile (since we linearly interpolate anything greater than 90th)
+        for i, t in enumerate(thresholds[:-1]):
```

Comment:
In this iteration, we take i to be the lower bound of each bin and i+1 (as you request below!) to be the upper bound of the bin (e.g. 01 to 10 percentile, 90 to 99 percentile). We need to keep 99 in there so we have a label for the upper value of bin 90-99. But we don't want i to ever equal 99 since there is no bin 99-?

Reply:
I've changed the code to add a list called "lower_thresholds" that makes it more clear what exactly we are doing.